substrate/frame/nomination-pools/src/lib.rs

```diff
@@ -3263,19 +3352,29 @@ impl<T: Config> Pallet<T> {
 }
 
 impl<T: Config> sp_staking::OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	/// Reduces the balances of the [`SubPools::with_era`], that belong to the pool involved in the
+	/// slash, to the amount that is defined in the `slashed_unlocking` field of
+	/// [`sp_staking::OnStakingUpdate::on_slash`]
+	///
+	/// Emits [`Event::<T>::PoolsSlashed`].
 	fn on_slash(
 		pool_account: &T::AccountId,
 		// Bonded balance is always read directly from staking, therefore we don't need to update
 		// anything here.
 		slashed_bonded: BalanceOf<T>,
 		slashed_unlocking: &BTreeMap<EraIndex, BalanceOf<T>>,
+		total_slashed: BalanceOf<T>,
 	) {
-		if let Some(pool_id) = ReversePoolIdLookup::<T>::get(pool_account) {
-			let mut sub_pools = match SubPoolsStorage::<T>::get(pool_id).defensive() {
-				Some(sub_pools) => sub_pools,
-				None => return,
-			};
-			for (era, slashed_balance) in slashed_unlocking.iter() {
+		let Some(pool_id) = ReversePoolIdLookup::<T>::get(pool_account).defensive() else { return };
+		// As the slashed account belongs to a [`BondedPool`] the [`TotalValueLocked`] decreases and
+		// an event is emitted.
+		TotalValueLocked::<T>::mutate(|tvl| {
+			tvl.saturating_reduce(total_slashed);
+		});
+
+		if let Some(mut sub_pools) = SubPoolsStorage::<T>::get(pool_id) {
```

Comment:
```suggestion
		if let Some(mut sub_pools) = SubPoolsStorage::<T>::get(pool_id).defensive() {
```

Reply:
Would you mind to tell me why this change should be introduced? I don't see why I would want to log an error when no `SubPool`'s are found.

Reply:
I could see this making sense if the param  `slashed_unlocking` would be changed to `Option(&BTreeMap)`

Reply:
I will put this in the next if let statement that tries to get the  `SubPool::with_era` as specified in the param `slashed_unlocking` because this one is actually always expected to be found.

Reply:
I think Kian meant that SubPools should always exist since we have already checked above that pool exists and there is some unlocking balance that has been slashed.

Reply:
Consider this setup: 
A pool gets created and there will be an active bond. Now a slash happens towards the validator that this pool selected. 
There should only be active stake for this pool as no unbonding has happenend.
The slashing event/hook will contain the pool's accountID and enter the `on_slash` fn. 

I think now we would expect the regular `BondedPool` to be found and have an event emitted but there should not be any defensive error about a SubPool not being found as it's completely legal to have no SubPool without any prior unbonding.

Reply:
> SubPools should always exist since we have already checked above that pool exists and there is some unlocking balance that has been slashed.

I think it's not guaranteed at this point that any unlocking balance has been slashed.

Reply:
Yes you are right, its not guaranteed if there has been no unbonding yet in the pool. Though if subpools don't exist, it would be good to assert `slashed_unlocking` is empty as well.

Reply:
Okay, I would suggest I change `slashed_unlocking` to be an `Option(BTreeMap)` and just pass `None` into the `on_slash`. This would seem very intuitive to me and cleaner to do these explicit checks. What do you think?

Reply:
Yes, good idea.